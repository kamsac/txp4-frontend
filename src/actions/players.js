import dotenvConfiguration from '../dotenv-configuration';
import PlayerResourceMock from '../resources/players/mock';
import PlayerResource from '../resources/players';

export const PLAYER_SUCCESS = 'PLAYER_SUCCESS';
export const PLAYER_REQUEST = 'PLAYER_REQUEST';
export const PLAYER_FAILURE = 'PLAYER_FAILURE';

export function requestPlayer() {
  return {
    type: PLAYER_REQUEST,
    payload: {
      isFetching: true,
    },
  };
}

export function receivePlayer(player) {
  return {
    type: PLAYER_SUCCESS,
    payload: {
      isFetching: false,
      player,
    },
  };
}

export function playerError() {
  return {
    type: PLAYER_FAILURE,
    payload: {
      isFetching: false,
    },
  };
}

export function loadPlayer(playerLogin) {
  return (dispatch) => {
    dispatch(requestPlayer());

    const resource = dotenvConfiguration.API_URL ? PlayerResource : PlayerResourceMock;

    resource.getPlayer(playerLogin)
      .then((response) => {
        dispatch(receivePlayer(response.data.player));
      })
      .catch(() => {
        dispatch(playerError());
      });
  };
}
