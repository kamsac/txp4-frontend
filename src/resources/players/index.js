import axios from 'axios';
import dotenvConfiguration from '../../dotenv-configuration';

const PlayerResource = {
  getPlayer(playerLogin) {
    return axios.get(`${dotenvConfiguration.API_URL}/players/${playerLogin}`);
  },

  getInventory(playerLogin) {
    return axios.get(`${dotenvConfiguration.API_URL}/players/${playerLogin}/inventory`);
  },

  equipItem(playerLogin, itemId) {
    const accessToken = localStorage.getItem('accessToken');
    return axios.get(`${dotenvConfiguration.API_URL}/players/${playerLogin}/inventory/${itemId}/equip`, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    });
  },

  unequipItem(playerLogin, itemId) {
    const accessToken = localStorage.getItem('accessToken');
    return axios.get(`${dotenvConfiguration.API_URL}/players/${playerLogin}/inventory/${itemId}/unequip`, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    });
  },
};

export default PlayerResource;
