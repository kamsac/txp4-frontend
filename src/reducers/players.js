import inventoryReducer from './inventory';
import { PLAYER_FAILURE, PLAYER_REQUEST, PLAYER_SUCCESS } from '../actions/players';
import {
  INVENTORY_REQUEST,
  INVENTORY_FAILURE,
  INVENTORY_SUCCESS,
  EQUIP_REQUEST,
  EQUIP_SUCCESS,
  EQUIP_FAILURE,
} from '../actions/inventory';

const initialState = {
  isFetching: false,
  items: {},
};

export default function playersReducer(state = initialState, action) {
  switch (action.type) {
    case PLAYER_REQUEST: {
      return Object.assign({}, state, {
        isFetching: action.payload.isFetching,
      });
    }
    case PLAYER_SUCCESS: {
      const { isFetching, player } = action.payload;
      return Object.assign({}, state, {
        isFetching,
        items: Object.assign({}, state.items, {
          [player.login]: {
            login: player.login,
            nick: player.nick,
            inventory: {
              isFetching: false,
              value: player.inventory,
            },
          },
        }),
      });
    }
    case PLAYER_FAILURE: {
      const { isFetching } = action.payload;
      return Object.assign({}, state, {
        isFetching,
      });
    }

    case INVENTORY_REQUEST:
    case INVENTORY_SUCCESS:
    case INVENTORY_FAILURE:
    case EQUIP_REQUEST:
    case EQUIP_SUCCESS:
    case EQUIP_FAILURE: {
      const { playerLogin } = action.payload;

      const nextState = Object.assign({}, state);

      if (!nextState.items[playerLogin]) {
        nextState.items[playerLogin] = {};
      }

      const inventory = inventoryReducer(nextState.items[playerLogin].inventory, action);
      nextState.items[playerLogin].inventory = inventory;
      return nextState;
    }
    default:
      return state;
  }
}

export function getPlayer(state, playerLogin) {
  if (
    state.players &&
    state.players.items
  ) {
    return state.players.items[playerLogin];
  }

  return null;
}

export function getPlayerNick(state, playerLogin) {
  if (
    state.players &&
    state.players.items &&
    state.players.items[playerLogin]
  ) {
    return state.players.items[playerLogin].nick;
  }

  return null;
}
