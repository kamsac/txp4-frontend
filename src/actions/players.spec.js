import { receivePlayer } from './players';

test('receivePlayer', () => {
  const somePlayer = {
    login: 'someLogin',
    nick: 'someNick',
    inventory: {
      items: [],
    },
  };
  expect(receivePlayer(somePlayer)).toMatchSnapshot();
});
