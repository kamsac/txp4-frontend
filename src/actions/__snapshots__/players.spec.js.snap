// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`receivePlayer 1`] = `
Object {
  "payload": Object {
    "isFetching": false,
    "player": Object {
      "inventory": Object {
        "items": Array [],
      },
      "login": "someLogin",
      "nick": "someNick",
    },
  },
  "type": "PLAYER_SUCCESS",
}
`;
